### PR TITLE
refactor: use `useSyncExternalStore`

### PR DESCRIPTION
**Pull request by SukkaW**
The PR changes `useVal`'s implementation to use `useSyncExternalStore` introduced in React 18.

To be backward compatible with React 16 and React 17, a package `use-sync-external-store` (published by the React Team as a polyfill) is also introduced.

All test cases pass on my machine locally.

## Additional Context

**What is `useSyncExternalStore`?**

- https://github.com/reactwg/react-18/discussions/70
- https://github.com/reactwg/react-18/discussions/86

In short, with the new `useSyncExternalStore` hook, maintainers of state management libraries no longer need to use `forceUpdate`/`setState`. Instead, React can now subscribe to the external store, track its updates, and schedule a re-render with the newest value to avoid in-consistent UI being committed to the DOM.

See also:

- https://github.com/nanostores/react/issues/6
- https://github.com/nanostores/react/pull/7

### CODE REVIEW

File: src/use-val.ts [hunk 1]
```diff
@@ -1,6 +1,11 @@
 import { type ReadonlyVal } from "value-enhancer";
 
-import { useDebugValue, useEffect, useState } from "react";
+import { useCallback, useDebugValue } from "react";
+import { useSyncExternalStore } from "use-sync-external-store";
```
Comment (hyrious) [inline src/use-val.ts]: Shouldn't it be `from 'use-sync-external-store/shim/index.js'`?

**Comment by crimx** [comment]
`useSyncExternalStore` does not handle the case where the val value is the same but it is actually a different version. In this case we still want a re-rendering be triggered for it.

For example, a `ReactiveMap` or `ReactiveSet` triggers changes when adding/updating/deleting values. The subscribers get the same `ReactiveMap`/`ReactiveSet` instance every time.

File: src/use-val.ts (hunk 1 above)
Comment (SukkaW) [reply src/use-val.ts]: Oh right, it should be `use-sync-external-store/shim`.

**Comment by SukkaW** [comment]
> For example, a `ReactiveMap` or `ReactiveSet` triggers changes when adding/updating/deleting values. The subscribers get the same `ReactiveMap`/`ReactiveSet` instance every time.

Oh, I see. That's what the `$version` property is for.

Given that `useSyncExternalStore` uses referential equality to determine if a re-render is necessary, let me explore if I can accomplish this without doing some "immutable" (deep clone) stuff.

**Review by crimx — APPROVED** [review]
